### PR TITLE
Bugfix: torch.multinomial samples same elements multiple times when replacement=False

### LINKED ISSUE #25030: torch.multinomial samples same elements multiple times when replacement=False
## 🐛 Bug

Sometimes a sampled index is drawn multiple time in a row even if replacement=False on `torch.multinomial`. This happens when an element of `input` is very small but not zero.

## To Reproduce

Here is a minimal working example:
```python 
import torch
torch.manual_seed(0)
w = torch.tensor( [1.2899e-01, 6.2532e-01, 3.6483e-02, 1.5196e-01, 2.9675e-03, 
4.9773e-03,4.5881e-02, 2.9019e-03, 5.2139e-04, 1.5281e-17] )
print(torch.multinomial(w, 10, replacement=False))
```
.

## Expected behavior

Raise an error because one element of `input` is almost zero.

## Environment

PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.2

Versions of relevant libraries:
[pip3] numpy==1.16.4
[pip3] torch==1.2.0
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.3.0
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @vincentqb @fritzo @neerajprad @alicanb @vishwakftw @nikitaved

### PR DESCRIPTION
**Pull request by ifedan**
Fixes https://github.com/pytorch/pytorch/issues/25030

Fixes torch.multinomial samples from sampling same elements multiple times when replacement=False

The issue exists because we lose mantissa when calculating cumulative distribution, summarizing large and small numbers, and distribution is not longer monotonically increasing.

The approach, that I prepose, is to use two sum aggregators, one will aggregate small numbers and other one will aggregate large numbers. Instead of row normalization, the uniform random sample will be multiplied by the row's two sums. And then after each sample we will recompute again the cumulative distribution.

### CODE REVIEW

**Comment by gchanan** [comment]
Summary of the technique here?

**Comment by ifedan** [comment]
> Summary of the technique here?

The issue exists because we lose mantissa when calculating cumulative distribution, summarizing large and small numbers, and distribution is not longer monotonically increasing.

The approach, that I prepose, is to use two sum aggregators, one will aggregate small numbers and other one will aggregate large numbers. Instead of row normalization, the uniform random sample will be multiplied by the row's two sums. And then after each sample we will recompute again the cumulative distribution.

**Comment by zou3519** [comment]
> > Summary of the technique here?
> 
> The issue exists because we lose mantissa when calculating cumulative distribution, summarizing large and small numbers, and distribution is not longer monotonically increasing.
> 
> The approach, that I prepose, is to use two sum aggregators, one will aggregate small numbers and other one will aggregate large numbers. Instead of row normalization, the uniform random sample will be multiplied by the row's two sums. And then after each sample we will recompute again the cumulative distribution.

(I haven't read the PR at all, so here are some superficial comments) It would be good to toss this into the commit message (and pull request body). I imagine it is hard to test this; how did you attempt to write a test for it?

**Comment by ifedan** [comment]
There are already some tests for multinomial exist.
Because the problem is related to cumulative distribution, I'm checking that this distribution is increasing:
https://github.com/pytorch/pytorch/pull/27523/files#diff-3abfe59c815ba32dd3d464e412fa0693R164

I also added test to cover cases that were introduced in #25030

**Review by zou3519 — CHANGES_REQUESTED** [review]
@ifedan and I discussed the following offline.

This PR introduces the following to fix multinomial without replacement:
- Recompute the CDF every time we have made a sample.
- Use two sum aggregators to fix precision problems
- Use accscalar_t to fix precision problems.
- Use two branches; one branch performs the original behavior and the other branch uses additional memory equal to the size of the probabilities tensor to recompute the CDF every time we sample. There is a warning in the branch that uses additional memory.

Below is our analysis on each individual point.

Recomputing the CDF is the key to correct multinomial-without-replacement behavior. We agree that this we should keep this approach.

Using two sum aggregators to fix precision problems
- The two sum aggregators still result in precision issues. One of the aggregators sums `floor(value)` for all values and the other one sums `frac(value)` for all values. A distribution that looks like [1e-17, 1e-34] doesn't work very well with these aggregators.
- In fact, the precision issue that the two aggregators tries to solve is not too bad. If we have a distribution that looks like [1e17, 1e-17], we would expect index `1` to get sampled with probability 1e-34 (which is effectively never in my books). 
- Because the two sum aggregators don't really help precision in all cases, and adds more complexity to the algorithm, we think it would be better to stick with one sum.

Use accscalar_t to fix precision problems
- This is probably good to have, especially when we are computing the cumulative sum, to avoid precision issues that arise from adding a lot of numbers together. Iirc in torch.sum we accumulate into accscalar_t; it is nice to do that here as well.

Using two branches
- We think that it is cleaner to just use the one branch that performs the correct behavior.
- The single branch may use additional memory equal to the size of the probabilities that records which indices have already been sampled. I'm not 100% sure that we need this extra memory, but we don't think users will miss it especially because it may be necessary for correctness.

tl;dr: action items are:
- Don't use two sum aggregators; use a single aggregator
- Don't use two branches; try to keep the implementation simple with one branch

but we're also curious what other people think.

File: aten/src/ATen/native/cpu/MultinomialKernel.cpp [hunk 1]
```diff
@@ -36,85 +37,189 @@ void multinomial_apply(Tensor& result, const Tensor& self, const int64_t n_sampl
...
-        if (cum_prob < uniform_sample) {
-          left_pointer = mid_pointer + 1;
-        }
-        else {
-          right_pointer = mid_pointer;
+      for (int64_t j = 0; j < n_sample; j++) {
+        /* sample a probability mass from a uniform distribution */
+        at::uniform_real_distribution<double> uniform(0, 1);
+        double uniform_sample = uniform(gen);
+        /* Do a binary search for the slot in which the prob falls
```
Comment (zou3519) [inline aten/src/ATen/native/cpu/MultinomialKernel.cpp]: BTW, it would be nice to use std::lower_bound so you don't have to reimplement binary search

**Comment by ifedan** [comment]
I have another idea: we lose precision when calculating sum, using addition, but we will not lose it if we will use multiplication instead:
cumsum = \sum_{i=0}^{N}Xi
Xi =ln(Yi) => Yi = e^{Xi}
\sum_{i=0}^{N}ln(Yi) = ln\prod_{i=0}^{N}Yi
\sum_{i=0}^{N}Xi = ln\prod_{i=0}^{N}e^{Xi}

**Comment by ifedan** [comment]
> I have another idea: we lose precision when calculating sum, using addition, but we will not lose it if we will use multiplication instead:
> image
> image
> image
> image



> I have another idea: we lose precision when calculating sum, using addition, but we will not lose it if we will use multiplication instead:
> image
> image
> image
> image

@zou3519 @gchanan what do you think?

**Comment by ifedan** [comment]
https://github.com/pytorch/pytorch/issues/28390

File: aten/src/ATen/native/cpu/MultinomialKernel.cpp [hunk 2]
```diff
@@ -94,27 +90,12 @@ void multinomial_apply(Tensor& result, const Tensor& self, const int64_t n_sampl
...
-          }
-          /* make total marginals sum to one */
-          new_val /= sum;
-          cum_dist_ptr[k * cum_dist_stride_0] = new_val;
+        self_copy_ptr[sample_idx] = 0;
+        sum = 0;
+        for (int64_t j = 0; j < n_categories; j++) {
+          val = self_copy_ptr[j];
+          sum += val;
+          cum_dist_ptr[j * cum_dist_stride_0] = sum;
```
Comment (andreaskoepf) [inline aten/src/ATen/native/cpu/MultinomialKernel.cpp]: Maybe the variables
```
auto cum_dist_stride_0 = cum_dist.stride(0);
auto self_copy_stride_0 = self_copy.stride(0);
``` 
should be be removed? I think there are no cases in which newly created tensors could be non-contiguous, are there? At least in ln 93 `self_copy_ptr` is indexed without it, so this should also work for the 1d `cum_dist` tensor.

**Comment by andreaskoepf** [comment]
I think it is a reasonable solution to follow the Cuda implementation and re-compute the cumulative sum always when an element has been drawn and its probability ise set to zero.

@ifedan regarding your exp idea: This would give smaller entries a much higher precision than larger ones. It would also require a normalization (with a sum computed in advance) for any larger vector. Otherwise the product could quickly become too larger (inf), since for all elements x > 0 -> exp(x) > 1, if the probabilities are not normalized. 

e.g.
```
>>> torch.rand(250).exp().prod().log()
tensor(inf)
```

**Review by zou3519 — CHANGES_REQUESTED** [review]
I haven't looked at the code yet, but could you provide some before / after numbers? Using accscalar will and should result in slower code, but I'm curious to see how much slower it is.

It would be great to also compare the performance of this to numpy's multinomial code (their api is called numpy.random.choice) to do some quick sanity checks.

File: aten/src/ATen/native/cpu/MultinomialKernel.cpp (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/native/cpu/MultinomialKernel.cpp]: Agreed

File: aten/src/ATen/native/cpu/MultinomialKernel.cpp [hunk 3]
```diff
@@ -123,7 +104,8 @@ void multinomial_apply(Tensor& result, const Tensor& self, const int64_t n_sampl
 
 static void multinomial_kernel_impl(Tensor& result, const Tensor& self, const int64_t n_sample, const bool with_replacement, Generator *gen) {
   AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "multinomial", [&] {
-    multinomial_apply<scalar_t>(result, self, n_sample, with_replacement, gen);
+    using accscalar_t = at::acc_type<scalar_t, false>;
```
Comment (zou3519) [inline aten/src/ATen/native/cpu/MultinomialKernel.cpp]: For boolean arguments it is nice to include a comment for what they are so that they're easier to read. i.e. `at::acc_type<scalar_t, /*is_cuda=*/false>`

File: test/test_torch.py [hunk 4]
```diff
@@ -9554,6 +9554,28 @@ def test_cdist_norm_batch(self, device):
...
+            # ensure [0.01 - 1.01] range
+            large_part = torch.rand(part_size, dtype=dtype, device=device) + 0.01
+            small_part = torch.rand(part_size, dtype=dtype, device=device) + 0.01
+            large_part = large_part * 1e15
+            small_part = small_part * 1e-15
+            dist = torch.cat((large_part, small_part), 0)
+            result = torch.multinomial(dist, num_samples, replacement=False)
+            actual = torch.unique(result, sorted=True)
+            expected = torch.from_numpy(np.arange(0, num_samples)).to(device)
+            self.assertEqual(actual, expected, "non-unique indicies")
```
Comment (zou3519) [inline test/test_torch.py]: "non-unique indices" is not very descriptive (i.e. what has non-unique indices? Why is it a problem that there are non-unique indices?), could we improve it?

File: test/test_torch.py [hunk 5]
```diff
@@ -9554,6 +9554,28 @@ def test_cdist_norm_batch(self, device):
...
+
+            dist = torch.cat((small_part, large_part), 0)
+            result = torch.multinomial(dist, num_samples, replacement=False)
```
Comment (zou3519) [inline test/test_torch.py]: It looks like we can refactor this into some sort of a helper function:

```
def check_multinomial(dist, num_samples):
    result = torch.multinomial(dist, num_samples, replacement=False)
    actual = ...
    expected = ...
    self.assertEqual(...)

```

File: aten/src/ATen/native/cpu/MultinomialKernel.cpp (hunk 2 above)
Comment (zou3519) [reply aten/src/ATen/native/cpu/MultinomialKernel.cpp]: If we really want to be certain, we can construct the tensor and then assert it is contiguous right after the construction.

File: test/test_torch.py [hunk 6]
```diff
@@ -9554,6 +9554,28 @@ def test_cdist_norm_batch(self, device):
                         expected = brute_cdist(x, y, p=p)
                         self.assertTrue(torch.allclose(expected, actual))
 
+    def test_multinomial_unnormalized(self, device):
+        for dtype in [torch.float, torch.double]:
```
Comment (zou3519) [inline test/test_torch.py]: It would probably be nice to check for CUDA half as well? But that test might (and I hope) exists already

File: aten/src/ATen/native/cpu/MultinomialKernel.cpp [hunk 7]
```diff
@@ -55,27 +61,17 @@ void multinomial_apply(Tensor& result, const Tensor& self, const int64_t n_sampl
...
-      for (int64_t j = 0; j < n_categories; j++) {
-        cum_dist_ptr[j * cum_dist_stride_0] /= sum;
-      }
-    }
-
     for (int64_t j = 0; j < n_sample; j++) {
       /* sample a probability mass from a uniform distribution */
       at::uniform_real_distribution<double> uniform(0, 1);
-      double uniform_sample = uniform(gen);
+      double uniform_sample = uniform(gen) * sum;
```
Comment (zou3519) [inline aten/src/ATen/native/cpu/MultinomialKernel.cpp]: Why did we stop normalizing the cumulative probability distribution? Does this make it more numerically stable?

File: aten/src/ATen/native/cpu/MultinomialKernel.cpp [hunk 8]
```diff
@@ -20,27 +21,32 @@ void multinomial_apply(Tensor& result, const Tensor& self, const int64_t n_sampl
   int64_t n_dist = self.dim() > 1 ? self.size(-2) : 1;
 
   /* cumulative probability distribution vector */
-  Tensor cum_dist = at::empty({n_categories}, self.options());
+  Tensor cum_dist = at::empty({n_categories}, self.options().dtype<accscalar_t>());
+  /* copy of original distribution to recalculate cdf */
+  Tensor self_copy = at::empty({n_categories}, self.options().dtype<scalar_t>());
```
Comment (zou3519) [inline aten/src/ATen/native/cpu/MultinomialKernel.cpp]: Two comments here:
- For the `with_replacement=True` case, we don't need `self_copy`. Let's not try to use too much extra memory when unnecessary.
- We don't really need a full copy of self, we just need a bitmap of whether or not an element has already been chosen. It is fine to assume that the user has at most 64 dimensions and use a std::bitmap<64> for this check. TensorIteratorReduce makes this assumption as well; there should be a global constant somewhere for this number (64)

File: aten/src/ATen/native/cpu/MultinomialKernel.cpp [hunk 9]
```diff
@@ -55,27 +61,17 @@ void multinomial_apply(Tensor& result, const Tensor& self, const int64_t n_sampl
...
       int left_pointer = 0;
       int right_pointer = n_categories;
       int mid_pointer;
-      scalar_t cum_prob;
+      accscalar_t cum_prob;
       int sample_idx;
-      /* Make sure the last cumulative distribution bucket sums to 1 */
-      cum_dist_ptr[(n_categories - 1) * cum_dist_stride_0] = 1;
 
       while(right_pointer - left_pointer > 0) {
```
Comment (zou3519) [inline aten/src/ATen/native/cpu/MultinomialKernel.cpp]: Please use std::lower_bound or file an issue to improve this to use std::lower_bound

File: aten/src/ATen/native/cpu/MultinomialKernel.cpp [hunk 10]
```diff
@@ -94,27 +90,12 @@ void multinomial_apply(Tensor& result, const Tensor& self, const int64_t n_sampl
```
Comment (zou3519) [inline aten/src/ATen/native/cpu/MultinomialKernel.cpp]: (commenting here bc gh won't let me comment above) what does "(will be incremented for lua compat by wrapper)" mean?

Comment (andreaskoepf) [reply aten/src/ATen/native/cpu/MultinomialKernel.cpp]: I think the lua-comment can be safely removed. I guess it was related to the indicies, since first element in arrays in lua is at position 1...

File: aten/src/ATen/native/cpu/MultinomialKernel.cpp (hunk 8 above)
Comment (andreaskoepf) [reply aten/src/ATen/native/cpu/MultinomialKernel.cpp]: Did you mean std::bitset? A 64 element limit sounds strange to me - wouldn't that mean that only 64 classes are supported? That would be definitely too restrictive. A dynamically allocated bitmap could be used and the recalculation of the sum could be omitted by just summing up until the random threshhold-point hab been found ... but it would only be a small improvement.

Comment (zou3519) [reply aten/src/ATen/native/cpu/MultinomialKernel.cpp]: @andreaskoepf, I did mean std::bitset and you're completely right, I got `n_categories` and the number of dimensions mixed up. Sorry about that!

File: aten/src/ATen/native/cpu/MultinomialKernel.cpp (hunk 7 above)
Comment (andreaskoepf) [reply aten/src/ATen/native/cpu/MultinomialKernel.cpp]: The normalization would require N additional divisions per round (when with_replacement==False) which are not really necessary...

File: aten/src/ATen/native/cpu/MultinomialKernel.cpp (hunk 8 above)
Comment (zou3519) [reply aten/src/ATen/native/cpu/MultinomialKernel.cpp]: Using a dynamically allocated bit array (`std::vector<bool>`) would save some memory. i.e., if we have n_categories = 1024 and we're working with float tensors, then instead of 4 * 1024 = 4kb we'd only need 1024 * 1/8 = 128 bytes.

File: test/test_torch.py (hunk 4 above)
Comment (zou3519) [reply test/test_torch.py]: Also, I'm not sure that `message` is the third argument to `self.assertEqual` (I think precision is the third argument). Please pass it as a kwarg.

**Comment by andreaskoepf** [comment]
@zou3519 regarding new performance: In with_replacement == False I would expect the new version to be significantly slower than the old implementation: The reason is the full rebuild of all cumulative sum entries instead of only patching and correcting following elements... probably it is ~50% slower (takes 2x the time of the original impl).

But since the probabilities vector is completely traversed each time some memory could be saved by removing the vector for the cumulative sum elements and by not using binary search but linear search during the summation pass over the probabilities. In pseudo-code: ```total = sum(prob); while (need_samples) { x = uniform_rng() * total; total=0; found=-1; for (i=0; i<nelem; i++) { if (!bitmap[i]) continue; v = prob[i]; if (found <0 && total+v>= x) { bitmap[i] = false; found=i; } else { total += v; } } result.push(found); }``` ... It would require two distinct implementations for with_replacement True and False cases - therefore I am not sure it is worth it.

The only other option I see to get the original performance would be to convert the probabilities into int64 with some quirks, e.g. stretching the probabilities over a large interval, e.g. (0, numeric_limits<int64>::max/2) and ensuring all probs >0 would become at least 1. With integral types the original algorithm would work since the sum-corrections would always be right)...

**Comment by rgommers** [comment]
> It would be great to also compare the performance of this to numpy's multinomial code (their api is called numpy.random.choice) to do some quick sanity checks.

Drive-by comment on this: numpy 1.17 has a new random number generator infrastructure, and `numpy.random.funcname` will use the legacy generator. Use (e.g.) `rng = np.random.default_rng()` to get one of the new generators (`PCG-64`), and use the `multinomial` and `choice` methods on that - it's several times faster than the legacy code, and has better statistical properties.
